### ISSUE TITLE
'Remove unnecessary cast' adds unnecessary parentheses

### ISSUE THREAD

**Issue by Neme12** [dotnet/roslyn#25554]
Inside a switch case:
```c#
switch (true)
{
    case (bool)default(bool):
}
```
```c#
switch (true)
{
    case (default(bool)):
}
```

inside a when clause:
```c#
switch (true)
{
    case true when (bool)default(bool):
}
```
```c#
switch (true)
{
    case true when (default(bool)):
}
```

inside an is pattern expression:
```c#
if (true is (bool)default(bool)) ;
```
```c#
if (true is (default(bool))) ;
```

Tagging @CyrusNajmabadi who's working on a code fix specifically to remove unnecessary parentheses. The problem is inside `ParenthesizedExpressionSyntaxExtensions.CanRemoveParentheses`. The first two cases should be trivial to fix but that's not the case inside `is`, where associativity plays a role. For example it's important to add tests so that parentheses are *not* removed here:
```c#
if (true is (true == true)) ;
```

**Comment by CyrusNajmabadi** [comment]
> The first two cases should be trivial to fix but that's not the case inside is, where associativity plays a role. For example it's important to add tests so that parentheses are not removed here:

This should already be checked in my PR.  precedence is different here, so we would consider those parens necessary.

**Comment by CyrusNajmabadi** [comment]
> Tagging @CyrusNajmabadi who's working on a code fix specifically to remove unnecessary parentheses. The problem is inside ParenthesizedExpressionSyntaxExtensions.CanRemoveParentheses. The first two cases should be trivial to fix

I can do this if you'd like.  But i don't want to step on your toes if you wanted to take this.

**Comment by Neme12** [comment]
I was going to go after this but then realized you might have fixed the issue inside is to take care of precedence since you changed a lot in that method. I'd prefer if you took care of this and added tests for these cases to tests for your analyzer and not to 'Remove unnecessary cast'.

**Comment by Neme12** [comment]
So if I was going to do this I would potentially have to duplicate some of your logic and step on *your* toes.

**Comment by Neme12** [comment]
A fix for this will break a few tests I added though: https://github.com/dotnet/roslyn/pull/25538#discussion_r175212884 that currently expect these extra parens. Depending on which PR is going to get merged first, soebody will have to take care of that later.

**Comment by CyrusNajmabadi** [comment]
No prob.  I'll make the change.  Whoever goes in second will just do the fixup.

**Comment by CyrusNajmabadi** [comment]
@Neme12 Have made the changes here: https://github.com/dotnet/roslyn/pull/21316  Want to take a look?

**Comment by Neme12** [comment]
@CyrusNajmabadi My PR was merged so you might want to fix some of those tests in your branch.

**Comment by Neme12** [comment]
This is fixed in 15.8 preview 2. @CyrusNajmabadi I think you fixed this with #26523 right?
@jinujoseph This can be closed
